### PR TITLE
[server] fix using snapshotBucketId when it is undefined or empty

### PR DESCRIPTION
**Pull request by sagor999**
## Related Issue(s)

Fixes https://github.com/gitpod-io/gitpod/issues/11823

## Release Notes

```release-note
none
```

## Werft options:

- [x] /werft with-preview

### CODE REVIEW

**Comment by sagor999** [comment]
@kylos101 FYI :eyes:

File: components/server/src/workspace/workspace-starter.ts [hunk 1]
```diff
@@ -1424,7 +1424,7 @@ export class WorkspaceStarter {
         }
 
         const volumeSnapshotId =
-            SnapshotContext.is(workspace.context) || WithPrebuild.is(workspace.context)
+            ((SnapshotContext.is(workspace.context) || WithPrebuild.is(workspace.context)) && workspace.context.snapshotBucketId !== undefined && workspace.context.snapshotBucketId != "")
```
Comment (kylos101) [inline components/server/src/workspace/workspace-starter.ts:1427]: Hey @sagor999 , here's a brief suggestion to also cover null:

```suggestion
((SnapshotContext.is(workspace.context) || WithPrebuild.is(workspace.context)) && !!workspace.context.snapshotBucketId)
```

[Context](https://www.typescriptlang.org/play?#code/MYewdgziA2CmB0sBOSRIBQEJNgK7WgEoBuAKFEhgWVQ21zABNYAzASzFkZPPCjkQo0WTACJRPCv2pC6YqAFtYAAgAuACw4BzCWSA)

Comment (kylos101) [reply components/server/src/workspace/workspace-starter.ts:1427]: [Moar context](https://stackoverflow.com/a/7452731)

Comment (sagor999) [reply components/server/src/workspace/workspace-starter.ts:1427]: Thank you!! :heart:

**Comment by sagor999** [comment]
@geropl could you confirm is passing undefined into
`const volumeSnapshots = await this.workspaceDb.trace(traceCtx).findVolumeSnapshotById(volumeSnapshotId);`
which calls:
https://github.com/gitpod-io/gitpod/blob/14b392003764fa1883285a6c1a114998c9fd243f/components/gitpod-db/src/typeorm/workspace-db-impl.ts#L741-L744
`findOne(volumeSnapshotId)`
I suspect it does not handle undefined variable properly here. But maybe I am wrong.
Does it? I have a feeling it returns random record if undefined is passed into it.

**Comment by geropl** [comment]
> I suspect it does not handle undefined variable properly here. But maybe I am wrong.
> Does it? I have a feeling it returns random record if undefined is passed into it.

You're right in that it is not handled properly. `undefined` should map to an "empty" value, which that is depends on the column definition/mapping code. So it could be that it pulls any record that is empty (haven't checked nor tested).

**Review by geropl — APPROVED** [review]
Code LGTM! :+1:

**Pull request merged**